consensus.go

```diff
@@ -0,0 +1,240 @@
+	"fmt"
+	"sort"
+	"time"
+
+	"decred.org/dcrwallet/v2/errors"
+	w "decred.org/dcrwallet/v2/wallet"
+	"github.com/decred/dcrd/chaincfg/chainhash"
+	"github.com/decred/dcrd/chaincfg/v3"
+)
+
+type Consensus struct {
+	mwRef *MultiWallet
+	ctx   context.Context
+}
+
+func newConsensus(mwRef *MultiWallet) (*Consensus, error) {
+	c := &Consensus{
+		mwRef: mwRef,
+	}
+
+	return c, nil
+}
+
+// GetVoteChoices handles a getvotechoices request by returning configured vote
+// preferences for each agenda of the latest supported stake version for the
+// selected wallet.
+func (c *Consensus) GetVoteChoices(ctx context.Context, hash string, walletID int) (*GetVoteChoicesResult, error) {
+	wallet := c.mwRef.WalletWithID(walletID)
+	wal := wallet.Internal()
+	if wal == nil {
+		return nil, fmt.Errorf("request requires a wallet but wallet has not loaded yet")
+	}
+
+	var ticketHash *chainhash.Hash
+	if hash != "" {
+		hash, err := chainhash.NewHashFromStr(hash)
+		if err != nil {
+			return nil, fmt.Errorf("inavlid hash: %w", err)
+		}
+		ticketHash = hash
+	}
+
+	version, agendas := getAllAgendas(wal.ChainParams())
+	resp := &GetVoteChoicesResult{
+		Version: version,
+		Choices: make([]VoteChoice, len(agendas)),
+	}
+
+	choices, _, err := wal.AgendaChoices(ctx, ticketHash)
+	if err != nil {
+		return nil, err
+	}
+
+	for i := range choices {
+		resp.Choices[i] = VoteChoice{
+			AgendaID:          choices[i].AgendaID,
+			AgendaDescription: agendas[i].Vote.Description,
+			ChoiceID:          choices[i].ChoiceID,
+			ChoiceDescription: "", // Set below
+		}
+		for j := range agendas[i].Vote.Choices {
+			if choices[i].ChoiceID == agendas[i].Vote.Choices[j].Id {
+				resp.Choices[i].ChoiceDescription = agendas[i].Vote.Choices[j].Description
+				break
+			}
+		}
+	}
+
+	return resp, nil
+}
+
+// SetVoteChoice handles a setvotechoice request by modifying the preferred
+// choice for a voting agenda.
+//
+// If a VSP host is configured in the application settings, the voting
+// preferences will also be set with the VSP.
+func (c *Consensus) SetVoteChoice(walletID int, vspPubKey []byte, vspHost, agendaID, choiceID, hash, passphrase string) error {
+	wallet := c.mwRef.WalletWithID(walletID)
+	wal := wallet.Internal()
+	if wal == nil {
+		return fmt.Errorf("request requires a wallet but wallet has not loaded yet")
+	}
+
+	err := wallet.UnlockWallet([]byte(passphrase))
+	if err != nil {
+		return translateError(err)
+	}
+	defer wallet.LockWallet()
+
+	var ticketHash *chainhash.Hash
+	if hash != "" {
+		hash, err := chainhash.NewHashFromStr(hash)
+		if err != nil {
+			return fmt.Errorf("inavlid hash: %w", err)
+		}
+		ticketHash = hash
+	}
+
+	choice := w.AgendaChoice{
+		AgendaID: agendaID,
+		ChoiceID: choiceID,
+	}
+
+	ctx := wallet.shutdownContext()
+	_, err = wal.SetAgendaChoices(ctx, ticketHash, choice)
+	if err != nil {
+		return err
+	}
+
+	if vspHost == "" {
+		return fmt.Errorf("request requires a vspHost but no vspHost was provided")
+	}
```

Comment:
If this is required, make this check much earlier. Also ensure that the vspClient can be created and return any error early.

Reply:
performing this check earlier prevents the  agenda choices from being saved locally, that should come first before the vsp attempt

Reply:
performing this check earlier prevents the  agenda choices from being saved locally, that should come first before the vsp attempt

Reply:
Okay, so the vsp host isn't required. Perhaps this shouldn't return an error then? dcrwallet returns a nil error if there isn't a configured vsp host.

Reply:
the vsp host is required, the method executes after agendas are saved locally, so it's still a requirement

Reply:
If the vsp host is required, this method shouldn't do anything unless it is provided. It doesn't make much sense to save the choices locally, then return an error.

Also, if a ticket hash is provided, is the vsp host parameter also required? I expect that if a ticket hash is provided, then choices should be saved for only that ticket, not for any other ticket.

It looks to me that either a ticket hash or vsp host is required. Should validate early on that either is provided and valid, before saving the choice locally. Because an error response after saving the choice locally might give the user the impression that nothing was saved.

Reply:
alright that makes sense, a check for the vsp host would be made earlier

yes that's how it's currently setup, if a ticket hash is provided the choices are saved for only that ticket